src/operator/deinput.rs

```diff
@@ -0,0 +1,703 @@
+/// stream.
+///
+/// A trait for a type that wraps a [`CollectionHandle`] or
+/// an [`UpsertHandle`] and pushes serialized relational data to the
+/// associated input stream record-by-record.  The client passes a
+/// deserializer object that provides access to a serialized data
+/// record (e.g., in JSON or CSV format) to [`insert`](`Self::insert`)
+/// and [`delete`](`Self::delete`) methods.  The record gets deserialized
+/// into the strongly typed representation expected by the input stream
+/// and gets buffered inside the handle.  The [`flush`](`Self::flush`)
+/// method pushes all buffered data to the underlying
+/// [`CollectionHandle`] or [`UpsertHandle`].
+///
+/// The exact serialized type passed to `insert` and `delete` methods
+/// depends on the underlying handle. See [`DeZSetHandle`], [`DeSetHandle`],
+/// and [`DeMapHandle`] documentation for details.
+///
+/// This trait is object-safe, i.e., it can be converted to
+/// `dyn DeCollectionHandle` and invoked using dynamic dispatch.  This allows
+/// choosing the input format (encapsulated by the deserializer object) at
+/// runtime or even switching between different formats dynamically.  Object
+/// safety is achived by using the `erased_serde` crate that exposes an
+/// object-safe deserializer API ([`erased_serde::Deserializer`]), so that
+/// `DeCollectionHandle` methods don't need to be parameterized by the
+/// deserializer type.
+pub trait DeCollectionHandle: Send {
+    /// Buffer a new insert update.
+    ///
+    /// The `deserializer` argument wraps a single serialized record.
+    /// The value gets deserialized and pushed to the underlying input stream
+    /// handle as an insert update.
+    ///
+    /// Returns an error if deserialization fails, i.e., the serialized
+    /// representation is corrupted or does not match the value type of
+    /// the underlying input stream.
+    ///
+    /// See [`DeZSetHandle`], [`DeSetHandle`], and [`DeMapHandle`]
+    /// documentation for details.
+    fn insert(&mut self, deserializer: &mut dyn ErasedDeserializer) -> Result<(), EError>;
+
+    /// Buffer a new delete update.
+    ///
+    /// The `deserializer` argument wraps a single serialized record whose
+    /// type depends on the underlying input stream: streams created by
+    /// [`Circuit::add_input_zset`](`crate::Circuit::add_input_zset`)
+    /// and [`Circuit::add_input_set`](`crate::Circuit::add_input_set`)
+    /// methods support deletion by value, hence the serialized record must
+    /// match the value type of the stream.  Streams created with
+    /// [`Circuit::add_input_map`](`crate::Circuit::add_input_map`)
+    /// support deletion by key, so the serialized record must match the key
+    /// type of the stream.
+    ///
+    /// The record gets deserialized and pushed to the underlying input stream
+    /// handle as a delete update.
+    ///
+    /// Returns an error if deserialization fails, i.e., the serialized
```

Comment:
what happens if the deletion itself cannot be executed, i.e., the key does not exist?

Reply:
Deletion never fails.  It either creates a value with a negative weight (if it's a regular input stream) or gets ignored (if it's an upsert handle).